packages/core/index.d.ts

```diff
@@ -272,7 +272,7 @@ declare module '@rjsf/core' {
 
     export function withTheme<T = any>(
         themeProps: ThemeProps<T>,
-    ): React.ComponentClass<FormProps<T>> | React.StatelessComponent<FormProps<T>>;
+    ): React.ForwardRefExoticComponent<React.PropsWithoutRef<FormProps<T>> & React.RefAttributes<Form<T>>>;
```

Comment:
This typing isn't actually correct. 

the actual return type here is actually more like

<Props>React.ForwardRefExoticComponent<React.PropsWithoutRef<FormProps<Props>> & React.RefAttributes<Form<Props>>>;

however, there's no way to handle this with the type system and the types of React.ForwardRefExoticComponent. 

This will basically always make the <Form /> a <Form<any> /> component, forwarding an any to the formData prop (always). 

I suggested in #2279  (before seeing this pr) that returning typeof Form here might be a solution that aligns the resulting types more correctly with the  function here?

Reply:
Isn't `FormProps` generic with? Why not pass down the generic just like I did? Your proposal

> ```ts
> React.ForwardRefExoticComponent<React.PropsWithoutRef<FormProps> & React.RefAttributes<Form>>;
> ```

is just the same as my proposal, but without the generic?

Reply:
Btw, I also added the types test https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-jsonschema-form/react-jsonschema-form-tests.tsx to see the effect of my changes. Maybe you can point to the test cases where you think the typing is wrong?

Reply:
@zepatrik  will try to get a look at those tests in a little bit. 

yes, FormProps is generic the issue is that 

```
export function withTheme<T = any>(
        themeProps: ThemeProps<T>,
    ): React.ForwardRefExoticComponent<React.PropsWithoutRef<FormProps<T>> & React.RefAttributes<Form<T>>>;
```

this code uses `T` from the theme props as the value for the generic, when what's actually produced is a generic forward ref that takes some new type `FormDataType` (because I really don't like using T and U everywhere :-D) and a resulting `<FormDataType>React.ForwardRefExoticComponent<React.PropsWithoutRef<FormProps<FormDataType>> & React.RefAttributes<Form<FormDataType>>>` would be more correct, however to my knowledge, based on how ForwardRefExoticComponent is defined, it's impossible to return this from the withTheme function

The `Form` resulting from `withTheme` _should_ accept a generic, not re-use the generic from `withTheme`. 

returning `typeof Form` acts as though it's returning the generic class `class Form<FormDataType> extend React.Component<FormDataType>`, which maintains the ability to pass the formData generic into `<WithThemedForm<FormDataType> />`, which the current typings with ForwardRefExoticComponent cannot support